File: part01/AzureVeneziano.Netduino/AzureVeneziano.Netduino/MobileServiceClient.cs
using Cet.MicroJSON;
using Microsoft.SPOT;
using System;
using System.IO;
using System.Net;
using System.Text;

/*
 * Copyright by Mario Vernari, Cet Electronics
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
namespace AzureVeneziano.Netduino
{
    /// <summary>
    /// HTTP Azure-mobile service client 
    /// </summary>
    public class MobileServiceClient
    {
        public const string Read = "GET";
        public const string Create = "POST";
        public const string Update = "PATCH";

        private const string JsonMimeType = "application/json";


        /// <summary>
        /// Create a new client for HTTP Azure-mobile servicing
        /// </summary>
        /// <param name="serviceName">The name of the target service</param>
        /// <param name="applicationId">The application ID</param>
        /// <param name="masterKey">The access secret-key</param>
        public MobileServiceClient(
            string serviceName,
            string applicationId,
            string masterKey
            )
        {
            this.ServiceName = serviceName;
            this.ApplicationId = applicationId;
            this.MasterKey = masterKey;

            this._baseUri = "http://" + this.ServiceName + ".azure-mobile.net/";
        }


        private string _baseUri;

        public string ServiceName { get; private set; }
        public string ApplicationId { get; private set; }
        public string MasterKey { get; private set; }


        /// <summary>
        /// Operate a table-oriented function
        /// </summary>
        /// <param name="tableName">The target table exposed by the service</param>
        /// <param name="method">The HTTP method required</param>
        /// <param name="data">The data to upload, whereas required</param>
        /// <returns>The resulting JSON data</returns>
        public JToken TableOperation(
            string tableName,
            string method,
            JToken data = null
            )
        {
            var address = new Uri(
                this._baseUri + "tables/" + tableName
                );

            return this.OperateCore(
                address,
                method,
                data
                );
        }


        /// <summary>
        /// Operate an API-oriented function
        /// </summary>
        /// <param name="apiName">The target API exposed by the service</param>
        /// <param name="method">The HTTP method required (typically POST)</param>
        /// <param name="data">The data to upload, whereas required</param>
        /// <returns>The resulting JSON data</returns>
        public JToken ApiOperation(
            string apiName,
            string method,
            JToken data = null
            )
        {
            var address = new Uri(
                this._baseUri + "api/" + apiName
                );

            return this.OperateCore(
                address,
                method,
                data
                );
        }


        private JToken OperateCore(
            Uri uri,
            string method,
            JToken data
            )
        {
            //create a HTTP request
            using (var request = (HttpWebRequest)WebRequest.Create(uri))
            {
                //set-up headers
                var headers = new WebHeaderCollection();
                headers.Add("X-ZUMO-APPLICATION", this.ApplicationId);
                headers.Add("X-ZUMO-MASTER", this.MasterKey);

                request.Method = method;
                request.Headers = headers;
                request.Accept = JsonMimeType;

                if (data != null)
                {
                    //serialize the data to upload
                    string serialization = JsonHelpers.Serialize(data);
                    byte[] byteData = Encoding.UTF8.GetBytes(serialization);
                    request.ContentLength = byteData.Length;
                    request.ContentType = JsonMimeType;
                    request.UserAgent = "Micro Framework";
                    //Debug.Print(serialization);

                    using (Stream postStream = request.GetRequestStream())
                    {
                        postStream.Write(
                            byteData,
                            0,
                            byteData.Length
                            );
                    }
                }

                //wait for the response
                using (var response = (HttpWebResponse)request.GetResponse())
                using (var stream = response.GetResponseStream())
                using (var reader = new StreamReader(stream))
                {
                    //deserialize the received data
                    return JsonHelpers.Parse(
                        reader.ReadToEnd()
                        );
                };
            }
        }

    }
}

File: partLinkIt/AzureVenezianoLinkIt/MobileServiceClient.cpp
/*
* Copyright by Mario Vernari, Cet Electronics
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#if defined(ARDUINO) && ARDUINO >= 100
    #include "Arduino.h"
#else
    #include "WProgram.h"
#endif

#include <ArduinoJson.h>
#include <LWiFi.h>
#include <LWiFiClient.h>
#include <HttpClient.h>
#include "MobileServiceClient.h"


// default constructor
MobileServiceClient::MobileServiceClient(
    const char* serverName,
    const char* applicationId,
    const char* masterKey,
    const int serverPort
    ) :
    _serverName(serverName),
    _applicationId(applicationId),
    _masterKey(masterKey),
    _serverPort(serverPort),
    _receivedBody(NULL)
{
} //MobileServiceClient


// default destructor
MobileServiceClient::~MobileServiceClient()
{
    if (_receivedBody != NULL){
        delete [] _receivedBody;
        _receivedBody = NULL;
    }
} //~MobileServiceClient


int MobileServiceClient::tableOperation(
    const char* tableName,
    const char* method,
    JsonObject* data
    )
{
    char path[40];
    strcpy(path, "/tables/");
    strcat(path, tableName);

    return this->operateCore(
        &path[0],
        method,
        data
        );
}


int MobileServiceClient::apiOperation(
    const char* apiName,
    const char* method,
    JsonObject* data
    )
{
    char path[40];
    strcpy(path, "/api/");
    strcat(path, apiName);

    return this->operateCore(
        &path[0],
        method,
        data
        );
}
        
        
/*
Example of raw HTTP request:

POST /tables/TodoItem HTTP/1.1
Accept: application/json
X-ZUMO-APPLICATION: (app-id)
X-ZUMO-MASTER: (master-key)
Content-Type: application/json
User-Agent: Mozilla/5.0
Host: sharemydata.azure-mobile.net
Content-Length: 31
Connection: close

{"text":"ddd","complete":false}

*/
int MobileServiceClient::operateCore(
    const char* path,
    const char* method,
    JsonObject* data
    )
{
    if (_receivedBody != NULL){
        delete [] _receivedBody;
        _receivedBody = NULL;
    }

    //create a HTTP request
    LWiFiClient client;
    HttpClient http(client);

    //it's important to call this method, because we need to add headers and data to the request
    //without it, the request is meant finished as simply as calling "startRequest"
    http.beginRequest();

    int err = HTTP_ERROR_CONNECTION_FAILED;
    if (http.startRequest(_serverName, _serverPort, path, HTTP_METHOD_POST, HttpUserAgent) == HTTP_SUCCESS){
        //serialize the data to upload
        char buffer[1024];
        int length = data->printTo(buffer, sizeof(buffer));
        
        //set-up headers
        http.sendHeader("X-ZUMO-APPLICATION", _applicationId);
        http.sendHeader("X-ZUMO-MASTER", _masterKey);
        http.sendHeader("Accept", JsonMimeType);
        
        if (data == NULL){
            http.endRequest();
        }
        else{
            http.sendHeader("Content-Type", JsonMimeType);
            http.sendHeader("Content-Length", length);
            http.endRequest();
        
            client.write((uint8_t*)&buffer, length);
        }
        
        //wait for the response
        err = http.responseStatusCode();
        if (err >= 0)
        {
            // Usually you'd check that the response code is 200 or a
            // similar "success" code (200-299) before carrying on,
            // but we'll print out whatever response we get

            err = http.skipResponseHeaders();
            if (err >= 0)
            {
                // Now we've got to the body, so we can print it out
                unsigned long timeoutStart = millis();
                char c;
                int bodyLen = http.contentLength();
            
                if (bodyLen > 0){
                    _receivedBody = new char[bodyLen];
                    
                    // Whilst we haven't timed out & haven't reached the end of the body
                    while ( 
                        (http.connected() || http.available()) &&
                        ((millis() - timeoutStart) < kNetworkTimeout) 
                        )
                    {
                        if (http.available())
                        {
                            c = http.read();
                            // Print out this character
                            Serial.print(c);
                    
                            bodyLen--;
                            // We read something, reset the timeout counter
                            timeoutStart = millis();
                        }
                        else
                        {
                            // We haven't got any data, so let's pause to allow some to
                            // arrive
                            delay(kNetworkDelay);
                        }
                    }
                }
            }
        }
    }
    
    http.stop();
    return err;
}
